File: archive/pulumi-ts/Pr6262/lib/alb-stack.ts
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface AlbStackArgs {
  environmentSuffix: string;
  vpcId: pulumi.Output<string>;
  publicSubnetIds: pulumi.Output<string[]>;
  ecsServiceArn: pulumi.Output<string>;
  targetGroupArn: pulumi.Output<string>;
  blueTargetGroupArn: pulumi.Output<string>;
  greenTargetGroupArn: pulumi.Output<string>;
  tags: pulumi.Input<{ [key: string]: string }>;
}

export class AlbStack extends pulumi.ComponentResource {
  public readonly albDnsName: pulumi.Output<string>;
  public readonly albArn: pulumi.Output<string>;

  constructor(
    name: string,
    args: AlbStackArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:alb:AlbStack', name, args, opts);

    const {
      environmentSuffix,
      vpcId,
      publicSubnetIds,
      blueTargetGroupArn,
      tags,
    } = args;

    // ALB Security Group
    const albSecurityGroup = new aws.ec2.SecurityGroup(
      `payment-alb-sg-${environmentSuffix}`,
      {
        vpcId: vpcId,
        description: 'Security group for payment application load balancer',
        ingress: [
          {
            protocol: 'tcp',
            fromPort: 80,
            toPort: 80,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'HTTP from internet',
          },
          {
            protocol: 'tcp',
            fromPort: 443,
            toPort: 443,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'HTTPS from internet',
          },
        ],
        egress: [
          {
            protocol: '-1',
            fromPort: 0,
            toPort: 0,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow all outbound traffic',
          },
        ],
        tags: pulumi.all([tags]).apply(([t]) => ({
          ...t,
          Name: `payment-alb-sg-${environmentSuffix}`,
          Tier: 'Web',
        })),
      },
      { parent: this }
    );

    // Application Load Balancer
    const alb = new aws.lb.LoadBalancer(
      `payment-alb-${environmentSuffix}`,
      {
        name: `payment-alb-${environmentSuffix}`,
        loadBalancerType: 'application',
        subnets: publicSubnetIds,
        securityGroups: [albSecurityGroup.id],
        enableDeletionProtection: false,
        enableHttp2: true,
        enableCrossZoneLoadBalancing: true,
        tags: pulumi.all([tags]).apply(([t]) => ({
          ...t,
          Name: `payment-alb-${environmentSuffix}`,
        })),
      },
      { parent: this }
    );

    // ALB Listener
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const _listener = new aws.lb.Listener(
      `payment-alb-listener-${environmentSuffix}`,
      {
        loadBalancerArn: alb.arn,
        port: 80,
        protocol: 'HTTP',
        defaultActions: [
          {
            type: 'forward',
            targetGroupArn: blueTargetGroupArn,
          },
        ],
      },
      { parent: this }
    );

    // WAF Web ACL for SQL Injection and XSS Protection
    const sqlInjectionRuleSet = new aws.wafv2.WebAcl(
      `payment-waf-${environmentSuffix}`,
      {
        name: `payment-waf-${environmentSuffix}`,
        scope: 'REGIONAL',
        description:
          'WAF for payment application with SQL injection and XSS protection',
        defaultAction: {
          allow: {},
        },
        rules: [
          {
            name: 'AWSManagedRulesCommonRuleSet',
            priority: 1,
            overrideAction: {
              none: {},
            },
            statement: {
              managedRuleGroupStatement: {
                vendorName: 'AWS',
                name: 'AWSManagedRulesCommonRuleSet',
              },
            },
            visibilityConfig: {
              cloudwatchMetricsEnabled: true,
              metricName: 'AWSManagedRulesCommonRuleSetMetric',
              sampledRequestsEnabled: true,
            },
          },
          {
            name: 'AWSManagedRulesSQLiRuleSet',
            priority: 2,
            overrideAction: {
              none: {},
            },
            statement: {
              managedRuleGroupStatement: {
                vendorName: 'AWS',
                name: 'AWSManagedRulesSQLiRuleSet',
              },
            },
            visibilityConfig: {
              cloudwatchMetricsEnabled: true,
              metricName: 'AWSManagedRulesSQLiRuleSetMetric',
              sampledRequestsEnabled: true,
            },
          },
          {
            name: 'BlockSQLKeywords',
            priority: 3,
            action: {
              block: {},
            },
            statement: {
              orStatement: {
                statements: [
                  {
                    byteMatchStatement: {
                      fieldToMatch: {
                        queryString: {},
                      },
                      positionalConstraint: 'CONTAINS',
                      searchString: 'SELECT',
                      textTransformations: [
                        {
                          priority: 0,
                          type: 'LOWERCASE',
                        },
                      ],
                    },
                  },
                  {
                    byteMatchStatement: {
                      fieldToMatch: {
                        queryString: {},
                      },
                      positionalConstraint: 'CONTAINS',
                      searchString: 'INSERT',
                      textTransformations: [
                        {
                          priority: 0,
                          type: 'LOWERCASE',
                        },
                      ],
                    },
                  },
                  {
                    byteMatchStatement: {
                      fieldToMatch: {
                        queryString: {},
                      },
                      positionalConstraint: 'CONTAINS',
                      searchString: 'DROP',
                      textTransformations: [
                        {
                          priority: 0,
                          type: 'LOWERCASE',
                        },
                      ],
                    },
                  },
                ],
              },
            },
            visibilityConfig: {
              cloudwatchMetricsEnabled: true,
              metricName: 'BlockSQLKeywordsMetric',
              sampledRequestsEnabled: true,
            },
          },
        ],
        visibilityConfig: {
          cloudwatchMetricsEnabled: true,
          metricName: `payment-waf-${environmentSuffix}`,
          sampledRequestsEnabled: true,
        },
        tags: pulumi.all([tags]).apply(([t]) => ({
          ...t,
          Name: `payment-waf-${environmentSuffix}`,
        })),
      },
      { parent: this }
    );

    // Associate WAF with ALB
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const _wafAssociation = new aws.wafv2.WebAclAssociation(
      `payment-waf-assoc-${environmentSuffix}`,
      {
        resourceArn: alb.arn,
        webAclArn: sqlInjectionRuleSet.arn,
      },
      { parent: this }
    );

    // Outputs
    this.albDnsName = alb.dnsName;
    this.albArn = alb.arn;

    this.registerOutputs({
      albDnsName: this.albDnsName,
      albArn: this.albArn,
    });
  }
}

File: archive/pulumi-py/Pr7031/lib/alb_stack.py
"""
ALB Stack - Application Load Balancer with HTTPS Listener and S3 Logging
"""

from typing import Dict, List

import pulumi
import pulumi_aws as aws


class AlbStackArgs:
    """Arguments for AlbStack"""

    def __init__(
        self,
        environment_suffix: str,
        vpc_id: pulumi.Input[str],
        public_subnet_ids: List[pulumi.Input[str]],
        alb_sg_id: pulumi.Input[str],
        log_bucket_name: pulumi.Input[str],
        certificate_arn: pulumi.Input[str] = None,
        tags: Dict[str, str] = None
    ):
        self.environment_suffix = environment_suffix
        self.vpc_id = vpc_id
        self.public_subnet_ids = public_subnet_ids
        self.alb_sg_id = alb_sg_id
        self.log_bucket_name = log_bucket_name
        self.certificate_arn = certificate_arn
        self.tags = tags or {}


class AlbStack(pulumi.ComponentResource):
    """
    Application Load Balancer for loan processing application.
    """

    def __init__(
        self,
        name: str,
        args: AlbStackArgs,
        opts: pulumi.ResourceOptions = None
    ):
        super().__init__("custom:alb:AlbStack", name, {}, opts)

        self.environment_suffix = args.environment_suffix
        self.tags = args.tags
        self.certificate_arn = args.certificate_arn

        # Create Application Load Balancer
        self.alb = aws.lb.LoadBalancer(
            f"loan-alb-{self.environment_suffix}",
            name=f"loan-alb-{self.environment_suffix}",
            internal=False,
            load_balancer_type="application",
            security_groups=[args.alb_sg_id],
            subnets=args.public_subnet_ids,
            enable_deletion_protection=False,
            enable_http2=True,
            enable_cross_zone_load_balancing=True,
            access_logs=aws.lb.LoadBalancerAccessLogsArgs(
                bucket=args.log_bucket_name,
                enabled=True,
                prefix="alb-logs"
            ),
            tags={**self.tags, "Name": f"loan-alb-{self.environment_suffix}"},
            opts=pulumi.ResourceOptions(parent=self)
        )

        # Create Target Group
        self.target_group = aws.lb.TargetGroup(
            f"loan-tg-{self.environment_suffix}",
            name=f"loan-tg-{self.environment_suffix}",
            port=8080,
            protocol="HTTP",
            vpc_id=args.vpc_id,
            target_type="ip",
            deregistration_delay=30,
            health_check=aws.lb.TargetGroupHealthCheckArgs(
                enabled=True,
                healthy_threshold=2,
                interval=30,
                matcher="200",
                path="/",
                port="traffic-port",
                protocol="HTTP",
                timeout=5,
                unhealthy_threshold=3
            ),
            tags={**self.tags, "Name": f"loan-tg-{self.environment_suffix}"},
            opts=pulumi.ResourceOptions(parent=self)
        )

        # HTTP Listener
        if self.certificate_arn:
            # Redirect to HTTPS
            self.http_listener = aws.lb.Listener(
                f"loan-http-listener-{self.environment_suffix}",
                load_balancer_arn=self.alb.arn,
                port=80,
                protocol="HTTP",
                default_actions=[aws.lb.ListenerDefaultActionArgs(
                    type="redirect",
                    redirect=aws.lb.ListenerDefaultActionRedirectArgs(
                        port="443",
                        protocol="HTTPS",
                        status_code="HTTP_301"
                    )
                )],
                tags={**self.tags, "Name": f"loan-http-listener-{self.environment_suffix}"},
                opts=pulumi.ResourceOptions(parent=self, depends_on=[self.alb])
            )

            # HTTPS Listener
            self.https_listener = aws.lb.Listener(
                f"loan-https-listener-{self.environment_suffix}",
                load_balancer_arn=self.alb.arn,
                port=443,
                protocol="HTTPS",
                ssl_policy="ELBSecurityPolicy-TLS13-1-2-2021-06",
                certificate_arn=self.certificate_arn,
                default_actions=[aws.lb.ListenerDefaultActionArgs(
                    type="forward",
                    target_group_arn=self.target_group.arn
                )],
                tags={**self.tags, "Name": f"loan-https-listener-{self.environment_suffix}"},
                opts=pulumi.ResourceOptions(
                    parent=self,
                    depends_on=[self.alb, self.target_group]
                )
            )
        else:
            # Forward directly to target group
            self.http_listener = aws.lb.Listener(
                f"loan-http-listener-{self.environment_suffix}",
                load_balancer_arn=self.alb.arn,
                port=80,
                protocol="HTTP",
                default_actions=[aws.lb.ListenerDefaultActionArgs(
                    type="forward",
                    target_group_arn=self.target_group.arn
                )],
                tags={**self.tags, "Name": f"loan-http-listener-{self.environment_suffix}"},
                opts=pulumi.ResourceOptions(parent=self, depends_on=[self.alb])
            )

        self.register_outputs({
            "alb_arn": self.alb.arn,
            "alb_dns_name": self.alb.dns_name,
            "target_group_arn": self.target_group.arn
        })
